File: src/history/dps/house-robber-198.java
class Solution {
    public int rob(int[] nums) {
        //dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
        //dp[i + 1] = max(dp[i - 1] + nums[i + 1], dp[i])
        //...
        // temp = max(prev + nums[i], cur)
        // prev = cur
        //  cur = temp
        //return cur
        
        int temp = 0, cur = 0, prev = 0;
        
        for(int i = 0; i < nums.length; i++){
            temp = Math.max(prev + nums[i], cur);
            prev = cur;
            cur = temp;
        }
        
        return cur;
    }
}

File: src/dps/house-robber-198.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
          return 0
        dp = [0 for i in range(len(nums) + 1)]
        dp[1] = nums[0]
        for i in range(2, len(nums) + 1):
          dp[i] = max(dp[i - 1], dp[i - 2]  + nums[i - 1])
        return dp[-1]